src/main/java/gregtech/common/metatileentities/multi/multiblockpart/MetaTileEntityPassthroughHatchFluid.java

```diff
@@ -101,40 +132,62 @@ protected IItemHandlerModifiable createImportItemHandler() {
     }
 
     @Override
-    protected ModularUI createUI(EntityPlayer entityPlayer) {
+    public boolean usesMui2() {
+        return true;
+    }
+
+    @Override
+    public ModularPanel buildUI(PosGuiData guiData, PanelSyncManager guiSyncManager) {
         int rowSize = (int) Math.sqrt(getTier() + 1);
-        return createUITemplate(entityPlayer, rowSize)
-                .build(getHolder(), entityPlayer);
-    }
-
-    private ModularUI.Builder createUITemplate(EntityPlayer player, int rowSize) {
-        ModularUI.Builder builder = ModularUI.builder(GuiTextures.BACKGROUND, 176, 18 + 18 * rowSize + 94)
-                .label(6, 6, getMetaFullName());
-
-        for (int y = 0; y < rowSize; y++) {
-            for (int x = 0; x < rowSize; x++) {
-                int index = y * rowSize + x;
-                builder.widget(
-                        new TankWidget(fluidTankList.getTankAt(index), 89 - rowSize * 9 + x * 18, 18 + y * 18, 18, 18)
-                                .setBackgroundTexture(GuiTextures.FLUID_SLOT)
-                                .setContainerClicking(true, true)
-                                .setAlwaysShowFull(true));
+
+        int backgroundWidth = 9 * 18 + 14;
+        int backgroundHeight = 18 + 18 * rowSize + 94;
+
+        List<List<IWidget>> widgets = new ArrayList<>();
+        for (int i = 0; i < rowSize; i++) {
+            widgets.add(new ArrayList<>());
+            for (int j = 0; j < rowSize; j++) {
+                widgets.get(i).add(new GTFluidSlot().syncHandler(fluidTankList.getTankAt(i * rowSize + j))
+                        .background(GTGuiTextures.FLUID_SLOT));
             }
         }
-        return builder.bindPlayerInventory(player.inventory, GuiTextures.SLOT, 7, 18 + 18 * rowSize + 12);
+
+        BooleanSyncValue workingStateValue = new BooleanSyncValue(() -> workingEnabled, val -> workingEnabled = val);
+        return GTGuis.createPanel(this, backgroundWidth, backgroundHeight)
+                .child(IKey.lang(getMetaFullName()).asWidget().pos(5, 5))
+                .child(SlotGroupWidget.playerInventory().left(7).bottom(7))
+                .child(new Grid()
+                        .top(18).height(rowSize * 18)
+                        .minElementMargin(0, 0)
+                        .minColWidth(18).minRowHeight(18)
+                        .alignX(0.5f)
+                        .matrix(widgets))
+                .child(new ToggleButton()
+                        .top(18 * 2).left(18 * 8 + 7)
+                        .value(workingStateValue)
+                        .overlay(GTGuiTextures.BUTTON_FLUID_OUTPUT)
+                        .tooltipBuilder(t -> t.setAutoUpdate(true)
```

Comment:
minor nitpick, but this auto update call should be in the `tooltip(t -> ...)` method so it's only called once. not blocking though.
same with MetaTileEntityPassthroughHatchItem